c18n: Support for v8 garbage collector

dpgao opened the pull request:
This PR extends and exposes a few existing APIs so that the v8 garbage collector can function with c18n.

jrtc27 commented on lib/libc/gen/Symbol.map:
```diff
@@ -86,6 +86,7 @@ FBSD_1.0 {
 	dl_c18n_unwind_trusted_stack;
 	dl_c18n_is_trampoline;
 	dl_c18n_pop_trusted_stack;
+	dl_c18n_reflect_trampoline;
```
Also, what's the actual V8 use case for this? Given that we use it in elf_utils.c I could believe it has other uses, but I'm curious as to what specifically V8 is doing and whether this is the right API for it.

jrtc27 replied:
As for the second link that doesn't tell me what it's actually being used for. Neither does the first, really. Why can you not just use codeptr relocations to avoid the need to unwrap trampolines later?

dpgao commented on libexec/rtld-elf/rtld_c18n.c:
```diff
@@ -1112,7 +1131,13 @@ dl_c18n_pop_trusted_stack(struct dl_c18n_compart_state *state, void *tfs)
+	 * XXX Dapeng: We increased the size of struct dl_c18n_compart_state,
+	 * which broke the ABI for libunwind. Use cheri_getlen to workaround
+	 * this for now, while adding reserved space to future versions of
+	 * struct dl_c18n_compart_state.
```
Yes. It wasn't there before because the only consumer back then was libunwind and it didn't need the _range_ of each compartment's stack, just the _top_.

I'll try using symbol versioning.

dpgao commented:
@jrtc27 I've removed the commit that exposed `dl_c18n_reflect_trampoline`, removed the padding at the end of `struct dl_c18n_compart_state`, and introduced a new symbol version for the `dl_c18n_pop_trusted_stack` function, while keeping an implementation that's compatible with the older version.

The only other thing that I think will be added to `struct dl_c18n_compart_state` is the TLS pointer. Since the existing libunwind implementation will become incompatible anyway once c18n starts switching the TLS block in the future, I think it is fine that we break the ABI of `struct dl_c18n_compart_state` here.

People who upgrade to the next release with c18n turned on will temporarily use an out-of-date libunwind and the compatible version `dl_c18n_pop_trusted_stack` while RTLD is already switching TLS blocks. But I guess the chances of this causing any problem in the practice is very low.

jrtc27 commented on libexec/rtld-elf/rtld_c18n.c:
```diff
@@ -826,15 +826,15 @@ init_stk_table(struct stk_table *table, struct tcb_wrapper *wrap)
 	 */
-	static char dummy_stk;
-	sp = &dummy_stk;
+	static _Alignas(16) char dummy_stk[16];
```
But can't we just look at the Executive permission for the interrupted compartment's PC? I view it as a bug that RCSP is even inspected when not the correct stack pointer for the compartment in question.

dpgao replied:
Not necessarily. If we are interrupted within a trampoline, the PC will have the Executive permission but the interrupted compartment would be either the caller or the callee of the trampoline.

dpgao commented on libexec/rtld-elf/rtld_c18n.c:
```diff
@@ -1123,6 +1123,27 @@ dl_c18n_is_trampoline(uintptr_t pc, void *tfs)
+ * that ends at sp.
+ */
+__asm(".symver dl_c18n_pop_trusted_stack_v1,"
+	      "dl_c18n_pop_trusted_stack@FBSD_1.0, remove");
```
I wasn't aware of `__sym_compat` and was following the recommended practice by (https://maskray.me/blog/2020-11-26-all-about-symbol-versioning). But the `, remove` here shouldn't make a difference here so I'll switch to the macro.
